b2sdk/transfer/emerge/executor.py

```diff
@@ -326,61 +326,197 @@ def _find_unfinished_file_by_plan_id(
+                    if log_rejections:
+                        logger.debug(
+                            'Rejecting %s: part %s sha1 mismatch', file_.file_id, part.part_number
+                        )
+                    continue  # part.sha1 doesn't match - so we reupload
+
                 finished_parts[part.part_number] = part
+
             if finished_parts is None:
                 continue
+
             finished_parts_len = len(finished_parts)
-            if best_match_file is None or finished_parts_len > best_match_parts_len:
+
+            if finished_parts and (
+                best_match_file is None or finished_parts_len > best_match_parts_len
+            ):
                 best_match_file = file_
                 best_match_parts = finished_parts
                 best_match_parts_len = finished_parts_len
+
+            if eager_mode and best_match_file is not None:
+                break
+
         return best_match_file, best_match_parts
 
+    def _find_unfinished_file_by_plan_id(
+        self,
+        bucket_id,
+        file_name,
+        file_info,
+        emerge_parts_dict,
+        encryption: EncryptionSetting,
+        file_retention: Optional[FileRetentionSetting] = None,
+        legal_hold: Optional[LegalHold] = None,
+        custom_upload_timestamp: Optional[int] = None,
+        cache_control: Optional[str] = None,
+    ):
+        """
+        Search for a matching unfinished large file by plan_id in the specified bucket.
+
+        This function aims to locate a matching unfinished large file using the plan_id and the supplied parameters. 
+        It's used to resume an interrupted upload, centralizing the shared logic between `_find_unfinished_file_by_plan_id` 
+        and `_match_unfinished_file_if_possible`.
+
+        In case a matching file is found but has inconsistencies (for example, mismatching file info or encryption settings), 
+        the function checks if 'plan_id' is in file_info, as this is a prerequisite.
+
+        :param bucket_id: The identifier of the bucket where the unfinished file resides.
+        :param file_name: The name of the file to be matched.
+        :param file_info: Information about the file to be uploaded.
+        :param emerge_parts_dict: A dictionary containing the parts of the file to be emerged.
+        :param encryption: The encryption settings for the file.
+        :param file_retention: The retention settings for the file, if any.
+        :param legal_hold: The legal hold status of the file, if any.
+        :param custom_upload_timestamp: The custom timestamp for the upload, if any.
+        :param cache_control: The cache control settings for the file, if any.
+        
+        :return: A tuple of the best matching unfinished file and its finished parts. If no match is found, it returns `None`.
+
+        Note: This operation requires the application key to have 'listFiles' access.
```

Comment:
I am not convinced private methods should state permissions they require. In any case, there should be some consistency and currently there is none.

Reply:
This was my inspiration: https://github.com/Backblaze/b2-sdk-python/blob/1d2001ca2e4868feff23fbfe8867f0ef5d99bc09/b2sdk/transfer/emerge/executor.py#L412

But I agree, it is strange to state it like this for a private method. Also, the assert around 'plan_id' is weird, given this is a private method. Do you agree?

Reply:
I don't think there is anything wrong with the assert, but maybe a different exception class should be raised?